src/main/java/org/commcare/session/RemoteQuerySessionManager.java

```diff
@@ -129,4 +133,41 @@ public void populateItemSetChoices(QueryPrompt queryPrompt) {
         evalContextWithAnswers.setVariables(userAnswers);
         ItemSetUtils.populateDynamicChoices(queryPrompt.getItemsetBinding(), evalContextWithAnswers);
     }
+
+    // loops over query prompts and validates selection until all selections are valid
+    public void refreshItemSetChoices(Hashtable<String, String> userAnswers) {
+        OrderedHashtable<String, QueryPrompt> userInputDisplays = getNeededUserInputDisplays();
+        boolean dirty = true;
+        while (dirty) {
```

Comment:
Very minor point: It makes sense why this code should converge, since any invalid choices are blanked out, and the worst case scenario for 5 choice boxes is to loop through 5 times, blanking one answer each time until all are blank.

That being said, while(X) loops make for challenging debugging if anything goes wrong with the plan, since you'll get a hang-and-kill rather than an explicit error.

From a style/future proofing perspective, I think it might make sense to add an explicit fallback in here where you only loop over the questions N times (where N is the size of the question set) and throw a runtime exception if the dirty flag is still set after the Nth pass, rather than continuing to loop. That ensures that the code functioning unexpectedly will give you an explicit exception which we can follow up on, rather than a hang, which we may never be notified of from the crash logs.